assets/js/editor-components/product-stock-control/index.tsx

```diff
@@ -1,54 +1,67 @@
 /**
  * External dependencies
  */
-import CheckboxList from '@woocommerce/base-components/checkbox-list';
+import { __, sprintf } from '@wordpress/i18n';
 import { getSetting } from '@woocommerce/settings';
-import { useCallback, useState } from '@wordpress/element';
+import { useCallback, useState, useEffect } from '@wordpress/element';
+import { useDebounce } from 'use-debounce';
+import { ToggleControl } from '@wordpress/components';
 
 export interface ProductStockControlProps {
 	value: Array< string >;
 	setAttributes: ( attributes: Record< string, unknown > ) => void;
 }
 
+// Should out of stock items be hidden globally?
+const hideOutOfStockItems = getSetting( 'hideOutOfStockItems', false );
+
+// Get the stock status options.
+const allStockStatusOptions = getSetting( 'stockStatusOptions', {} );
+
 /**
  * A pre-configured SelectControl for product stock settings.
  */
 const ProductStockControl = ( {
 	value,
 	setAttributes,
 }: ProductStockControlProps ): JSX.Element => {
-	// Should out of stock items be hidden?
-	const [ hideOutOfStockItems ] = useState(
-		getSetting( 'hideOutOfStockItems', false )
-	);
-
-	// Get the stock status options.
-	const [ { outofstock, ...otherStockStatusOptions } ] = useState(
-		getSetting( 'stockStatusOptions', {} )
-	);
-
 	// Determine whether or not to use the out of stock status.
-	const [ STOCK_STATUS_OPTIONS ] = useState(
-		hideOutOfStockItems
-			? otherStockStatusOptions
-			: { outofstock, ...otherStockStatusOptions }
-	);
+	const { outofstock, ...otherStockStatusOptions } = allStockStatusOptions;
+	const stockStatusOptions = hideOutOfStockItems
+		? otherStockStatusOptions
+		: allStockStatusOptions;
 
 	// Set the initial state to the default or saved value.
 	const [ checkedOptions, setChecked ] = useState( value );
 
+	// Debounce checked options for updates.
+	const [ debouncedCheckedOptions ] = useDebounce< string[] >(
+		checkedOptions,
+		400
+	);
+
 	/**
 	 * Valid options must be in an array of [ 'value' : 'mystatus', 'label' : 'My label' ] format.
 	 * stockStatusOptions are returned as [ 'mystatus' : 'My label' ].
 	 * Formatting is corrected here.
 	 */
 	const [ displayOptions ] = useState(
-		Object.entries( STOCK_STATUS_OPTIONS )
+		Object.entries( stockStatusOptions )
 			.map( ( [ slug, name ] ) => ( { value: slug, label: name } ) )
 			.filter( ( status ) => !! status.label )
 			.sort( ( a, b ) => a.value.localeCompare( b.value ) )
 	);
 
+	/**
+	 * Dobounce changes to attributes based on checked items.
+	 */
+	useEffect( () => {
+		// The blank stock status prevents all results returning when all options are unchecked.
+		setAttributes( {
+			stockStatus: [ '', ...debouncedCheckedOptions ],
+		} );
+	}, [ debouncedCheckedOptions, setAttributes ] );
+
```

Comment:
Is this one fixing the hidden layout issue? ideally we shouldn't debounce attribute updates and let the editor handle those things.

Reply:
There was a state before this. It looks like some form of debounce was intended but I agree we should remove it.